src/test/integration/tests/basic.ts

```diff
@@ -344,4 +346,660 @@ export const tests: {[name: string] : SSRTest} = {
     ],
     stableSelectors: ['p'],
   },
+
+  'NodePart accepts nested templates': {
+    render(x: any, y: any) {
+      return html`<div>${x}${html`<span>${y}</span>`}</div>`;
+    },
+    expectations: [
+      {
+        args: ['A', 'B'],
+        html: '<div>A\n  <span>B</span></div>'
+      },
+      {
+        args: ['C', 'D'],
+        html: '<div>C\n  <span>D</span></div>'
+      }
+    ],
+    stableSelectors: ['div', 'span'],
+  },
+
+  'multiple NodeParts': {
+    render(x: any, y: any) {
+      return html`<div>${x}${y}</div>`;
+    },
+    expectations: [
+      {
+        args: ['A', 'B'],
+        html: '<div>A\n  B</div>'
+      },
+      {
+        args: ['C', 'D'],
+        html: '<div>C\n  D</div>'
+      }
+    ],
+    stableSelectors: ['div'],
+  },
+
+  'multiple NodeParts with whitespace': {
+    render(x: any, y: any) {
+      return html`<div>${x} ${y}</div>`;
+    },
+    expectations: [
+      {
+        args: ['A', 'B'],
+        html: '<div>A\n  B</div>',
+        check(assert: Chai.Assert, dom: HTMLElement) {
+          const childNodes = dom.querySelector('div')!.childNodes;
+          const textContent = filterNodes(childNodes, Node.TEXT_NODE)
+            .map(n => n.textContent);
+          assert.deepEqual(textContent, ['A', ' ', 'B']);
+        }
+      },
+      {
+        args: ['C', 'D'],
+        html: '<div>C\n  D</div>',
+        check(assert: Chai.Assert, dom: HTMLElement) {
+          const childNodes = dom.querySelector('div')!.childNodes;
+          const textContent = filterNodes(childNodes, Node.TEXT_NODE)
+            .map(n => n.textContent);
+          assert.deepEqual(textContent, ['C', ' ', 'D']);
+        }
+      }
+    ],
+    stableSelectors: ['div'],
+  },
+
+  'NodePart with trailing whitespace': {
+    render(x: any, y: any) {
+      return html`<div>${x} </div>`;
+    },
+    expectations: [
+      {
+        args: ['A'],
+        html: '<div>A\n  </div>',
+        check(assert: Chai.Assert, dom: HTMLElement) {
+          const childNodes = dom.querySelector('div')!.childNodes;
+          const textContent = filterNodes(childNodes, Node.TEXT_NODE)
+            .map(n => n.textContent);
+          assert.deepEqual(textContent, ['A', ' ']);
+        }
+      },
+      {
+        args: ['B'],
+        html: '<div>B\n  </div>',
+        check(assert: Chai.Assert, dom: HTMLElement) {
+          const childNodes = dom.querySelector('div')!.childNodes;
+          const textContent = filterNodes(childNodes, Node.TEXT_NODE)
+            .map(n => n.textContent);
+          assert.deepEqual(textContent, ['B', ' ']);
+        }
+      }
+    ],
+    stableSelectors: ['div'],
+  },
+
+  'AttributePart accepts a string': {
```

Comment:
Is there a suite concept? This starts attribute parts.

Reply:
We should add one soon. We should also split these into multiple files.

Reply:
For now I added comment blocks to delineate.